lib/Infer/InstSynthesis.cpp

```diff
@@ -0,0 +1,1264 @@
+    // don't subsequently wire the output to the input(s)
+    if (J == 0)
+      CompConstraint = getOutputLocVarConstraint(0, N, IC);
+    else {
+      CompConstraint = getOutputLocVarConstraint(N, N+J, IC);
+    }
+    // Init fresh loop PCs
+    auto LoopPCs = WiringPCs;
+    LoopPCs.emplace_back(CompConstraint, IC.getConst(APInt(1, true)));
+
+    // --------------------------------------------------------------------------
+    // -------------- Counterexample driven synthesis loop ----------------------
+    // --------------------------------------------------------------------------
+    unsigned Refinements = 0;
+    while (true) {
+      Inst *Query = IC.getConst(APInt(1, true));
+      // Put each set of concrete inputs into a separate copy of the WiringQuery
+      for (unsigned J = 0; J < S.size(); ++J) {
+        auto InputMap = S[J];
+        if (DebugSynthesis) {
+          for (auto const &Input : InputMap) {
+            if (Input.first->Name.find(COMP_INPUT_PREFIX) != std::string::npos)
+              continue;
+            llvm::outs() << "setting input " << Input.first->Name
+                         << " to " << Input.second->Val << "\n";
+          }
+        }
+        if (J > 0) {
+          // Starting with the second concrete input set,
+          // copy component input variables for query separation
+          for (auto const &E : LocInstMap) {
+            if (E.first.find(COMP_INPUT_PREFIX) != std::string::npos) {
+              auto In = E.second.second;
+              std::string Name = E.first + LOC_SEP + std::to_string(Refinements);
+              InputMap[In] = IC.createVar(In->Width, Name);
+            }
+          }
+        }
+        Inst *Copy = getInstCopy(WiringQuery, IC, InputMap);
+        Query = IC.getInst(Inst::And, 1, {Query, Copy});
+      }
+      // Solve the synthesis constraint.
+      // Each solution corresponds to a syntactically distinct and well-formed
+      // straight-line program obtained by composition of given components
+      std::vector<Inst *> ModelInsts;
+      std::vector<llvm::APInt> ModelVals;
+      InstMapping Mapping(Query, IC.getConst(APInt(1, true)));
+      // Negate the query to get a SAT model.
+      // Don't use original BPCs/PCs, they are useless
+      std::string QueryStr = BuildQuery({}, LoopPCs, Mapping,
+                                        &ModelInsts, /*Negate=*/true);
+      bool IsSat;
+      EC = SMTSolver->isSatisfiable(QueryStr, IsSat, ModelInsts.size(),
+                                    &ModelVals, Timeout);
+      if (EC)
+        return EC;
+
+      // No valid wiring exists for the target comp number
+      if (!IsSat)
+        break;
+
+      std::vector<std::pair<LocInst, LocInst>> Wiring;
+      Inst *Cand = createInstFromModel(std::make_pair(ModelInsts, ModelVals),
+                                       Wiring, IC);
+      if (!Cand) {
+        llvm::errs() << "synthesis bug: creating inst from a model failed\n";
```

Comment:
If this truly should never happen, I'd prefer to crash souper when it does, instead of printing something.